core/blockchain/blockchain_gateway.py

```diff
@@ -0,0 +1,355 @@
+        """
+        ipfs_hash = self.client.add_json(content)
+        return ipfs_hash
+
+    ##########################################################################
+    ###                         DEVELOPER SECTION                          ###
+    ##########################################################################
+
+    def broadcast_decentralized_learning(self, model_config: object)-> str:
+        """
+        Upload a model config and weights to the blockchain
+        """
+        tx_receipt = self.setter(None, model_config, True)
+        return tx_receipt
+
+    def broadcast_terminate(self, key: str) -> str:
+        """
+        Terminates decentralized training
+        TODO: check if training even started
+        """
+        tx_receipt = self.setter(key, None)
+        return tx_receipt
+
+    def handle_decentralized_learning_owner(self, model_config: object) -> None:
+        """
+        Return weights after training terminates
+        TODO: add condition to check if training for specific model terminated
+        """
+        final_weights = self.getter(header)
+        return final_weights
+
+    ##########################################################################
+    ###                          PROVIDER SECTION                          ###
+    ##########################################################################
+
+    async def start_listening(self, event_filter, handler, poll_interval=5):
+        """
+        Starts an indefinite loop that listens for a specific event to occur.
+        Called in `filter_set`. Filters are some condition that must be
+        fulfilled on a per tx basis
+        `poll_interval` specifies the number of seconds to stall before each
+        poll
+        """
+        while True:
+            filtered_diffs = self.get_state_diffs(event_filter)
+            if filtered_diffs:
+                return filtered_diffs
+            await asyncio.sleep(poll_interval)
+
+    def filter_set(self, event_filter, handler):
+        """
+        Calls async method `start_listening` and called by various listening
+        methods
+        """
+        asyncio.set_event_loop(asyncio.new_event_loop())
+        loop = asyncio.get_event_loop()
+        try:
+            filtered_diffs = loop.run_until_complete(
+                self.start_listening(event_filter, handler))
+            check = [handler(diff) for diff in filtered_diffs]
+        finally:
+            loop.close()
+        return check
+
+    def handle_decentralized_learning_trainer(self, tx: dict) -> None:
+        """
+        This callback will be triggered by the Listener if it finds the 
+        method signature it's looking for.
+        The parameters (model weights, model config) will be downloaded 
+        and put into the optimizer initially. So the optimizer knows this info.
+        """
+        logging.info("handling decentralized learning... {}".format(tx))
+        key = tx.get('key')
```

Comment:
Please check for errors while checking for these keys.

Reply:
there is no way a tx will not have 'key' because of how it is constructed

Reply:
@georgymh would `assert` statements help here?

Reply:
Yes